fix too restrictive method installations

ThomasBreuer opened the pull request:
In many situations,
we cannot expect that compatible groups have the same type.

resolves #4144

ThomasBreuer committed:
fix too restrictive method installations

In many situations,
we cannot expect that compatible groups have the same type.

ThomasBreuer committed:
fix signatures

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -1351,11 +1350,11 @@ function inner_automorphism_group(A::AutomorphismGroup{T}) where T <: GAPGroup
 Return whether `f`(`H`) == `H`.
 """
-function is_invariant(f::GAPGroupElem{AutomorphismGroup{T}}, H::T) where T<:GAPGroup
+function is_invariant(f::GAPGroupElem{<: AutomorphismGroup{<: GAPGroup}}, H::GAPGroup)
```
I think you also need to adjust `docs/src/Groups/autgroup.md` to match

ThomasBreuer committed:
next round

- adjust reference in .md file
- change more types in signatures
- extend tests
- fix `small_group` with given type: apply it to an Oscar group not a GAP group
- fix `sub`: the type check was wrong, and the compatibility is checked anyhow

lgoettgens commented on docs/src/Groups/autgroup.md:
````diff
@@ -11,8 +11,8 @@ automorphism_group(G::GAPGroup)
 ```@docs
-is_invariant(f::GAPGroupElem{AutomorphismGroup{T}}, H::T) where T<:GAPGroup
-restrict_automorphism(f::GAPGroupElem{AutomorphismGroup{T}}, H::T, A=automorphism_group(H)) where T <: GAPGroup
+is_invariant(f::GAPGroupElem{AutomorphismGroup{T}}, H::GAPGroup) where T <: GAPGroup
````
The T in this line is irrelevant for the first part

ThomasBreuer replied:
Hmm, I try to put the same line into the .md file that appears in the src file, and this seems to work.
Do we have a better convention (perhaps for the section on the `docs` folder in the developer documentation)?

lgoettgens replied:
This is in fact fine, I just didn't read properly. Sorry for the confusion

lgoettgens commented on docs/src/Groups/autgroup.md:
```diff
@@ -11,8 +11,8 @@ automorphism_group(G::GAPGroup)
-is_invariant(f::GAPGroupElem{AutomorphismGroup{T}}, H::T) where T<:GAPGroup
-restrict_automorphism(f::GAPGroupElem{AutomorphismGroup{T}}, H::T, A=automorphism_group(H)) where T <: GAPGroup
+is_invariant(f::GAPGroupElem{AutomorphismGroup{T}}, H::GAPGroup) where T <: GAPGroup
+restrict_automorphism(f::GAPGroupElem{AutomorphismGroup{T}}, H::GAPGroup, A=automorphism_group(H)) where T <: GAPGroup
```
Here as well

lgoettgens commented on src/Groups/homomorphisms.jl:
```diff
@@ -1298,7 +1297,7 @@ end
 #Base.:^(f::GAPGroupElem{AutomorphismGroup{T}},g::GAPGroupElem{AutomorphismGroup{T}}) where T <: GAPGroup = g^-1*f*g
 
-function (A::AutomorphismGroup{T})(f::GAPGroupHomomorphism{T,T}) where T <: GAPGroup
+function (A::AutomorphismGroup{S})(f::GAPGroupHomomorphism{T,T}) where S <: GAPGroup where T <: GAPGroup
```
```suggestion
function (A::AutomorphismGroup{S})(f::GAPGroupHomomorphism{T,T}) where {S <: GAPGroup, T <: GAPGroup}
```
a double `where` is almost never necessary.
Alternatively, you could even do
```suggestion
function (A::AutomorphismGroup{<: GAPGroup})(f::GAPGroupHomomorphism{T,T}) where T <: GAPGroup
```

ThomasBreuer replied:
o.k.
Concerning the multiple `where`, there are lots in many code files, this is perhaps something for a pull request of its own.

lgoettgens commented on src/Groups/libraries/smallgroups.jl:
```diff
@@ -83,7 +83,7 @@ Pc group of order 60
 """
 function small_group(::Type{T}, n::IntegerUnion, m::IntegerUnion) where T
-  G = _small_group(n, m)
+  G = small_group(n, m)
```
Is this change intended? I am not deep enough into the details to understand the difference

ThomasBreuer replied:
Yes. (It is also mentioned in the corresponding commit message.)

lgoettgens commented on src/Groups/sub.jl:
````diff
@@ -34,7 +34,6 @@ true
 ```
 """
 function sub(G::GAPGroup, gens::AbstractVector{S}; check::Bool = true) where S <: GAPGroupElem
````
```suggestion
function sub(G::GAPGroup, gens::AbstractVector{<: GAPGroupElem}; check::Bool = true)
```
since you don't use `S` anymore inside the function

ThomasBreuer replied:
o.k.

lgoettgens reviewed (approved):
looks good, just some minor comments

ThomasBreuer committed:
address comments (improve syntax)
